File: reverse_words.rb
def reverse_words(str)
  reverse_string(str)

  left = 0
  right = 1
  last_index = 0

  while last_index < str.length
    while str[right + 1] != ' ' && !str[right + 1].nil?
      right += 1
    end

    last_index = right + 2

    while left < right
      str[left], str[right] = str[right], str[left]
      left += 1
      right -= 1
    end

    left = last_index
    right = last_index + 1
  end

  str
end

def reverse_string(str)
  left = 0
  right = str.length - 1

  while left < str.length / 2
    str[left], str[right] = str[right], str[left]
    left += 1
    right -= 1
  end

  str
end

p reverse_words('the sky is blue')
p reverse_words('')
p reverse_words('hello')

File: reverse_words.js
var reverseWords = function(str) {
  var chars = reverseString(str.trim()).split('');
  var left = 0;
  var right = 1;
  var lastIndex = 0;

  while (right < chars.length) {
    while (chars[right + 1] !== ' ' && chars[right + 1] !== undefined) {
      right++;
    }

    lastCurrentWordIndex = right;

    while (chars[right] === ' ') {
      right++;
    }

    lastIndex = right;

    while (left < lastCurrentWordIndex) {
      var temp = chars[left];
      chars[left] = chars[lastCurrentWordIndex];
      chars[lastCurrentWordIndex] = temp;

      console.log(chars);
      console.log('left ' + left + ', right ' + right);
      left++;
      lastCurrentWordIndex--;
    }

    left = lastIndex;
    right = lastIndex + 1;
  }

  return chars.join('');
};

function reverseString(str) {
  var chars = str.split('');
  var left = 0;
  var right = str.length - 1;

  while (left < right) {
    var temp = chars[left];
    chars[left] = chars[right];
    chars[right] = temp;

    left++;
    right--;
  }

  return chars.join('');
}


console.log('\'' + reverseWords('hello    world') + '\'');
